bundles/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/internal/handler/EnergenieSocket.java

```diff
@@ -0,0 +1,186 @@
+        this.host = host;
+        key = getKey(password);
+    }
+
+    private byte[] getKey(final String password) {
+        final int passwordLength = password.length();
+        String passwordString = password;
+        for (int i = 0; i < (8 - passwordLength); i++) {
+            passwordString = passwordString + " ";
+        }
+        return passwordString.getBytes();
+    }
+
+    public synchronized byte[] sendCommand(final byte[] ctrl) throws IOException {
+        try (final TaskSocket taskSocket = authorize()) {
+            final OutputStream output = taskSocket.socket.getOutputStream();
+            final InputStream input = taskSocket.socket.getInputStream();
+
+            if (output == null || input == null) {
+                throw new IOException("No connection");
+            } else {
+                if (logger.isTraceEnabled()) {
+                    logger.trace("Control message send to EG (int) '{}' (hex)'{}'", ctrl, HexUtils.bytesToHex(ctrl));
+                }
+                output.write(encryptControls(ctrl, taskSocket.task));
+                readStatus(input, taskSocket);
+                return updateStatus(taskSocket);
+            }
+        }
+    }
+
+    public synchronized byte[] retrieveStatus() throws IOException {
+        try (final TaskSocket taskSocket = authorize()) {
+            return updateStatus(taskSocket);
+        }
+    }
+
+    private TaskSocket authorize() throws IOException {
+        final TaskSocket taskSocket = new TaskSocket();
+        final OutputStream output = taskSocket.socket.getOutputStream();
+        final InputStream input = taskSocket.socket.getInputStream();
+
+        if (output == null || input == null) {
+            throw new IOException("No connection");
+        }
+        output.write(MESSAGE);
+        logger.trace("Start Condition '{}' send to EG", MESSAGE);
+        input.read(taskSocket.task);
+
+        if (logger.isTraceEnabled()) {
+            logger.trace("EG responded with task (int) '{}' (hex) '{}'", taskSocket.task,
+                    HexUtils.bytesToHex(taskSocket.task));
+        }
+        final byte[] solutionMessage = calculateSolution(taskSocket.task);
+
+        output.write(solutionMessage);
+        logger.trace("Solution '{}' send to EG", solutionMessage);
+        readStatus(input, taskSocket);
+        return taskSocket;
+    }
+
+    private void readStatus(final InputStream input, final TaskSocket taskSocket) throws IOException {
+        input.read(taskSocket.statcryp);
+        if (logger.isTraceEnabled()) {
+            logger.trace("EG responded with statcryp (int) '{}' (hex) '{}'", taskSocket.statcryp,
+                    HexUtils.bytesToHex(taskSocket.statcryp));
+        }
+    }
+
+    private byte[] updateStatus(final TaskSocket taskSocket) throws IOException {
+        final InputStream input = taskSocket.socket.getInputStream();
+
+        if (input == null) {
+            throw new IOException("No connection");
```

Comment:
You aren't even using `input` in this method, why are you checking if it is null?

Reply:
see above

Reply:
removed